### PR TITLE
[#4] Implement long form of the identifiers

### PR DESCRIPTION
**Pull request by kgilden**
Alright, this PR adds support for long form of the identifiers.

* should be fully backwards compatible - all methods previously working with the first identifier octet also support passing a string of identifier octets;
* biggest supported identifier depends on PHP's integer size (either 32-bit or 64-bit). That's because I'm decoding long format identifiers to integers. Don't think this won't be a problem though as longer identifiers are probably very infrequent;
* consequently fixes a silent overflow in [ObjectIdentifier](https://github.com/FGrosse/PHPASN1/blob/1966cc2/lib/ASN1/Universal/ObjectIdentifier.php#L108) - an exception *should* be thrown instead;

If something feels off, feel free to point at it.

### CODE REVIEW

File: lib/ASN1/Base128.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
+<?php
```
Comment (kgilden) [inline lib/ASN1/Base128.php]: Really don't like having this under `FG\ASN1`. Would `FG\ASN1\Util` sound better? Or something entirely different?

File: lib/ASN1/Base128.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        $octets = chr($value & 0x7F);
+        $value >>= 7;
+
+        while ($value > 0) {
+            $octets .= chr(0x80 | ($value & 0x7F));
+            $value >>= 7;
+        }
+
+        return strrev($octets);
```
Comment (FGrosse) [inline lib/ASN1/Base128.php]: X.690 section 8.1.2.4.2 b) states the following about the encoding of the long form identifier:
> bits 7 to 1 of the first subsequent octet, followed by bits 7 to 1 of the second subsequent octet, followed in turn by bits 7 to 1 of each further octet, up to and including the last subsequent octet in the identifier octets shall be the encoding of an unsigned  binary integer equal to the tag number, with bit 7 of the first subsequent octet as the most significant bit;

This describes the octets following the first identifier octet which holds, the class, primitive/constructed, and long form marker. The important part here is at the very end of the paragraph: *with bit 7 of the first subsequent octet as the most significant bit*.

The way I understand this I would do the encoding as follows:

```
value:   0x80
binary:  10000000
septets:  1000000  0000000
padded:  11000000 00000000
hex:     0xC0     0x00
```

However this is not what is happening in your implementation and I am a bit uncertain about my interpretation of the standard and would be happy about your input.

I tried to come up with an alternative implementation but I just dont have enough time currently. Maybe I can finish it over the weekend but I can't promise anything. If I do I will push this to a separate branch and wil defer the merge until its completely clear to me how the encoding needs to be implemented.

**Comment by FGrosse** [comment]
Seems like you forgot to add the actual length of the identifier here:
https://github.com/FGrosse/PHPASN1/pull/33/files#diff-5312df74313165959d478b79ac37544aR163

File: lib/ASN1/Base128.php (hunk 2 above)
Comment (FGrosse) [inline lib/ASN1/Base128.php]: I compared your implementation with the one from three other DER encoding implementations (openssl, golang and ASN.1 JavaScript decoder) and they result in the same encoding so it seems like I misunderstood the standard about the encoding at this point.

From what I can reconstruct the correct encoding of `0x80`is:

```
value:   0x80
binary:  10000000
septets:  0000001  0000000
padded:  10000001 00000000
hex:     0x81     0x00
```

It actually does make more sense this way so feel free to ignore this commet block.
Sometimes it just helps writting down those things to get a better understanding of the whole thing :)

Comment (kgilden) [inline lib/ASN1/Base128.php]: @FGrosse Yeah, I ran into similar issues, until I read [this article](http://www.strozhevsky.com/free_docs/asn1_in_simple_words.pdf) (p3). So I connected the dots and figured both long identifiers and SID's are encoded the same way (base-128). And your second encoding seems the correct one to me as well.

**Comment by kgilden** [comment]
@FGrosse Good catch, fixed it!

File: lib/ASN1/Base128.php (hunk 2 above)
Comment (FGrosse) [inline lib/ASN1/Base128.php]: Your PR is scheduled for this evening as well and thanks for the article :)

**Comment by FGrosse** [comment]
So I finally merged this. Just got hung up in reading specs :)
As always big thanks for the pull request and happy contributing in the future

**Pull request merged**